File: medium/String/reverse-words-in-a-string/reverse_words.cpp
#include <string>
#include <vector>
#include <sstream>
#include <algorithm>
using namespace std;

class Solution {
public:
    string reverseWords(string s) {
        vector<string> words;
        stringstream ss(s);
        string word;
        
        while (ss >> word) {
            words.push_back(word);
        }
        
        reverse(words.begin(), words.end());
        
        string res;
        for (int i = 0; i < words.size(); ++i) {
            if (i > 0) res += " ";
            res += words[i];
        }
        return res;
    }
};

// Test
#include <iostream>
int main() {
    Solution sol;
    cout << sol.reverseWords("the sky is blue") << endl;
    return 0;
}

File: medium/String/reverse-words-in-a-string/reverse_words.py
def reverseWords(s: str) -> str:
    """
    Reverse the words in a string, remove extra spaces.
    Time: O(n)
    Space: O(n)
    """
    words = s.strip().split()
    words.reverse()
    return " ".join(words)

if __name__ == "__main__":
    print(reverseWords("the sky is blue"))  # "blue is sky the"
